File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/L.Фибоначчи_по_модулю/solution.cpp
#include <iostream>
#include <cmath>

int solution(long long n, long long k) {
    /*
    Time Complexity O(N)
    Memory Complexity O(1)
    */
    long long bound = std::pow(10, k);
    long long x = 1, y = 1;
    for (int i = 1; i < n; ++i) {
        long long c = y;
        y = (x % bound + y % bound) % bound;
        x = c;
    }
    return y;
}

int main() {
    long long n, k;
    std::cin >> n >> k;
    std::cout << solution(n, k) << "\n";
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/L.Фибоначчи_по_модулю/solution.py
import sys


def solution(n: int, k: int) -> int:
    """
    Time Complexity O(N)
    Memory Complexity O(1)
    """
    x = y = 1
    bound: int = 10**k
    for _ in range(1, n):
        x, y = y % bound, (x + y) % bound
    return y % bound


if __name__ == '__main__':
    n, k = map(int, sys.stdin.readline().rstrip().split())
    sys.stdout.write(str(solution(n, k)) + '\n')

